### ISSUE TITLE
Unity environment freezing on launch

### ISSUE THREAD

**Issue by AndreTInfante** [Unity-Technologies/ml-agents#13]
I've got to the point that I can successfully launch the 3D ball example. However, when I run the jupyter notebook file and the app starts automatically, it starts a tiny window and then freezes, and has to be process killed (see image below). Jupyter notebook doesn't show any errors, so I'm not entirely sure what's up. Any ideas as to how to fix this?  

![image](https://user-images.githubusercontent.com/3455125/30683844-2e151c34-9e6c-11e7-9f0e-936dd0fdc9a4.png)

Thank you!

**Comment by eagleEggs** [comment]
The small window is expected as it's set in the scene to train at that window size (128x72). Likely to make performance better while it's crunching the PPO. I have noticed on my system that it seems to freeze, while still actually crunching away. Check Tensorflow to see if it continues to update, also Jupyter updates the mean reward below the training code chunk as it progresses. Also, if it turns out to be resource related, maybe remove some of those balls/platforms from the scene and then rebuild.

Overall, this demo is pretty intensive and I'm guessing that watching the game window while it's crunching probably isn't feasible unless you're on a decent rig (I have 8GB RAM I7 and still can barely watch the game window during this specific PPO), but even then watching it is just wasting resources :/

**Comment by JZito** [comment]
Same problem but am able to launch Unity Environments successfully by launching the python scripts directly.

Following the wiki Balance Balls walkthrough, launching the environments from the Basics or PPO Jupyter Notebooks, Unity Environment will launch, display the Built by Unity logo and then hang with a black screen. 
This is after commenting out signal-related content from the environment.py file, as referenced here: https://github.com/Unity-Technologies/ml-agents/issues/5
Here is the output log:

Mono path[0] = 'C:/Documents/Unity/ml-agents/python/test2_Data/Managed'
Mono config path = 'C:/Documents/Unity/ml-agents/python/test2_Data/MonoBleedingEdge/etc'
PlayerConnection initialized from C:/Documents/Unity/ml-agents/python/test2_Data (debug = 0)
PlayerConnection initialized network socket : 0.0.0.0 55248
Multi-casting "[IP] 192.168.1.130 [Port] 55248 [Flags] 2 [Guid] 3478566033 [EditorId] 2953059523 [Version] 1048832 [Id] WindowsPlayer(DESKTOP-NQ9QNAK) [Debug] 0" to [225.0.0.222:54997]...
Started listening to [0.0.0.0:55248]
PlayerConnection already initialized - listening to [0.0.0.0:55248]
Player data archive not found at `C:/Documents/Unity/ml-agents/python/test2_Data/data.unity3d`, using local filesystem
Initialize engine version: 2017.1.1f1 (5d30cf096e79)
GfxDevice: creating device client; threaded=1
Direct3D:
    Version:  Direct3D 11.0 [level 11.0]
    Renderer: NVIDIA GeForce GTX 960M (ID=0x139b)
    Vendor:   NVIDIA
    VRAM:     4064 MB
    Driver:   21.21.13.7633
Begin MonoManager ReloadAssembly
Loading C:\Documents\Unity\ml-agents\python\test2_Data\Managed\UnityEngine.dll into Unity Child Domain
Platform assembly: C:\Documents\Unity\ml-agents\python\test2_Data\Managed\System.Core.dll (this message is harmless)
Loading C:\Documents\Unity\ml-agents\python\test2_Data\Managed\Assembly-CSharp.dll into Unity Child Domain
Loading C:\Documents\Unity\ml-agents\python\test2_Data\Managed\UnityEngine.UI.dll into Unity Child Domain
Loading C:\Documents\Unity\ml-agents\python\test2_Data\Managed\UnityEngine.Networking.dll into Unity Child Domain
Loading C:\Documents\Unity\ml-agents\python\test2_Data\Managed\UnityEngine.Timeline.dll into Unity Child Domain
Loading C:\Documents\Unity\ml-agents\python\test2_Data\Managed\Newtonsoft.Json.dll into Unity Child Domain
Platform assembly: C:\Documents\Unity\ml-agents\python\test2_Data\Managed\System.dll (this message is harmless)
- Completed reload, in  0.048 seconds
<RI> Initializing input.

<RI> Input initialized.

<RI> Initialized touch support.

UnloadTime: 1.216441 ms
Platform assembly: C:\Documents\Unity\ml-agents\python\test2_Data\Managed\System.Configuration.dll (this message is harmless)
Platform assembly: C:\Documents\Unity\ml-agents\python\test2_Data\Managed\System.Xml.dll (this message is harmless)
Platform assembly: C:\Documents\Unity\ml-agents\python\test2_Data\Managed\Mono.Security.dll (this message is harmless)
Platform assembly: C:\Documents\Unity\ml-agents\python\test2_Data\Managed\System.Data.dll (this message is harmless)

Once you force quit the program, the log is filled with cascading socket exceptions:

SocketException: An established connection was aborted by the software in your host machine.

I was able to successfully launch the environment running the python script directly with "python ppo.py <env_name> --train".

**Comment by awjuliani** [comment]
Thanks for trying this out, everyone. 

It is indeed the case that running in a small window, and low frame-rate are expected behavior. This is in order to allow the game to run at 100x speed, and provide as much training data as possible. If this is creating issues on your device, I would recommend trying to decrease the `Time Scale` under `Training Configuration` of the Academy object in the scene to something like `10`. See https://github.com/Unity-Technologies/ml-agents/wiki/Agents-Editor-Interface#academy for more details. While this will slow training, it will also provide more visual feedback that the environment is running as expected. 

The black screen hang that you are experiencing @JZito might be due to the environment not resetting yet. Try running a block of code in `Basic.ipynb` which runs `env.reset()`, and see if the screen is still black.

**Comment by AndreTInfante** [comment]
Hmm. I slowed down the time scale to 5 for debugging purpose, and I could see it going. But after a second or two the pixels on screen stopped moving and it froze. I was watching it in jupyter notebook, and it was showing new rewards until the freeze, at which point it stopped.

**Comment by AndreTInfante** [comment]
Nevermind, I'm an idiot, I needed to increase the number of episodes. Thanks for your help!

**Comment by awjuliani** [comment]
Glad you were able to get things working!

**Comment by gautam1858** [comment]
@awjuliani Can you please tell me how to reduce the time scale ?

**Comment by quantumiracle** [comment]
Hi,
I also meet this problem that launching from the python script or jupyter notebook as [here](https://github.com/Unity-Technologies/ml-agents/blob/master/notebooks/getting-started-gym.ipynb) for a self-built env will stuck at a small screen. And it seems it stucks before or at the env.reset() step. I've made the timescale to be 1 for training, but it still stucks. Do you guys know what's wrong with it?

**Comment by jialululu** [comment]
Hi there,
we also came across similar problems. The code hangs after we run the env.reset() block. and on the small window, we could still control the angle we see using our mouses. The whole program just gets stuck on the reset part. Does anyone have any idea what kind of error we are running into?

**Comment by awjuliani** [comment]
Hi all. To have the simulation proceeds from python, you must call `env.step()` from python. Unity will wait otherwise for the action from python. If you'd like to run in a simple loop, you can call `env.step()` without passing any actions. Otherwise, you would compute actions for the agents in your scene to take.
